src/topology/config/mod.rs

```diff
@@ -127,19 +127,37 @@ pub enum DataType {
 
 #[typetag::serde(tag = "type")]
 pub trait SourceConfig: core::fmt::Debug {
-    fn build(
-        &self,
-        name: &str,
-        globals: &GlobalOptions,
-        shutdown: ShutdownSignal,
-        out: mpsc::Sender<Event>,
-    ) -> crate::Result<sources::Source>;
+    fn build(&self, cx: SourceContext) -> crate::Result<sources::Source>;
 
     fn output_type(&self) -> DataType;
 
     fn source_type(&self) -> &'static str;
 }
 
+#[derive(Debug, Clone)]
+pub struct SourceContext<'a> {
+    pub name: &'a str,
+    pub globals: &'a GlobalOptions,
+    pub shutdown: ShutdownSignal,
+    pub out: mpsc::Sender<Event>,
+}
+
+impl<'a> SourceContext<'a> {
+    pub fn new_test(
```

Comment:
#[cfg(test)]?

Reply:
Oops, forgot to add it! Other contexts have it, I'll be sure to add it here too.
Good catch!

Reply:
Actually, `TransformContext::new_test` isn't `#[cfg(test)]` either. This is odd, maybe we should add it there too?

Reply:
Nah, tests fail if we add `#[cfg(test)]` to `TransformContext::new_test`. It doesn't mean we don't want to do it - it's just not that simple - so not for this PR.